dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/converter/RelationshipTrackerConverterService.java

```diff
@@ -0,0 +1,245 @@
+import org.hisp.dhis.relationship.RelationshipType;
+import org.hisp.dhis.trackedentity.TrackedEntityInstance;
+import org.hisp.dhis.tracker.TrackerIdentifier;
+import org.hisp.dhis.tracker.domain.Enrollment;
+import org.hisp.dhis.tracker.domain.Event;
+import org.hisp.dhis.tracker.domain.Relationship;
+import org.hisp.dhis.tracker.domain.RelationshipItem;
+import org.hisp.dhis.tracker.domain.TrackedEntity;
+import org.hisp.dhis.tracker.preheat.TrackerPreheat;
+import org.hisp.dhis.tracker.preheat.TrackerPreheatParams;
+import org.hisp.dhis.tracker.preheat.TrackerPreheatService;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Date;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static org.hisp.dhis.relationship.RelationshipEntity.PROGRAM_INSTANCE;
+import static org.hisp.dhis.relationship.RelationshipEntity.PROGRAM_STAGE_INSTANCE;
+import static org.hisp.dhis.relationship.RelationshipEntity.TRACKED_ENTITY_INSTANCE;
+
+/**
+ * @author Enrico Colasante
+ */
+@Service
+public class RelationshipTrackerConverterService
+    implements TrackerConverterService<Relationship, org.hisp.dhis.relationship.Relationship>
+{
+    private final TrackerPreheatService trackerPreheatService;
+
+    private final TrackerConverterService<TrackedEntity, TrackedEntityInstance> trackedEntityTrackerConverterService;
+
+    private final TrackerConverterService<Enrollment, ProgramInstance> enrollmentTrackerConverterService;
+
+    private final TrackerConverterService<Event, ProgramStageInstance> eventTrackerConverterService;
+
+    public RelationshipTrackerConverterService( TrackerPreheatService trackerPreheatService,
+        TrackerConverterService<Enrollment, ProgramInstance> enrollmentTrackerConverterService,
+        TrackerConverterService<Event, ProgramStageInstance> eventTrackerConverterService,
+        TrackerConverterService<TrackedEntity, TrackedEntityInstance> trackedEntityTrackerConverterService )
+    {
+        this.trackerPreheatService = trackerPreheatService;
+        this.enrollmentTrackerConverterService = enrollmentTrackerConverterService;
+        this.eventTrackerConverterService = eventTrackerConverterService;
+        this.trackedEntityTrackerConverterService = trackedEntityTrackerConverterService;
+    }
+
+    @Override
+    @Transactional( readOnly = true )
+    public Relationship to( org.hisp.dhis.relationship.Relationship relationship )
+    {
+        List<Relationship> relationships = to( Collections.singletonList( relationship ) );
+
+        if ( relationships.isEmpty() )
+        {
+            return null;
+        }
+
+        return relationships.get( 0 );
+    }
+
+    @Override
+    @Transactional( readOnly = true )
+    public List<Relationship> to( List<org.hisp.dhis.relationship.Relationship> relationships )
```

Comment:
This class converts objects from one type to another, do we need to run the conversion in a database TX? Can't we cache the lookup data and run outside a tx?

Reply:
Transactional here was completely wrong.
Removing it